### DISCUSSION TITLE
Block default action

### DISCUSSION THREAD

**Discussion by pintassilgo** [sezanzeb/input-remapper#443] [Q&A]
I need to remap extra mouse button (forward) + scroll to change audio volume. Done it like this:
![image](https://user-images.githubusercontent.com/5483864/180915676-46b686c1-8b5b-4227-8080-2abe511395d3.png)

But there's one issue: default actions are not being prevented and that way it's unusable for me.

If I press extra button + roll the wheel, volume changes but hovered page/app also scrolls. That scrolling shouldn't happen.

Similarly, forward button is also acting on combination when suitable, but should be prevented.

In other words:

Expected result:

- mouse forward button + scroll up → just raise audio volume, nothing else, just like if I pressed the multimedia key from my keyboard

Actual result:

- mouse forward button + scroll up → default forward button action + scroll up + raise audio volume

Is there any way to block default action of remapped keys/buttons in a combination?

Thanks.

--- \*\* **Edit** \*\* ---

I just tested Beta, the recorded input adds "Wheel HI RES". That plus "Release Input" enabled (default) is causing scroll to be cancelled. That fixes part of the issue.
![image](https://user-images.githubusercontent.com/5483864/180928994-57c12e6e-782b-480f-a38d-764238c94d4e.png)

But forward action (like going forward one page in browser tab history) is still being executed when it shouldn't. I believe the reason is as follows:
Forward action is triggered on "button down" (hold), not on "button up" (release). So when I trigger the secondary input of the combination (wheel), forward action already ran because forward button was already "clicked".

So I need to disable forward action on "button down" (hold) and enable it on "button up" (release), because I'm using it as a modifier. That way I believe its default action could be cancelled when I click a secondary button in combination (in this case, turning mouse wheel). How to do that with input-remapper?

I'm able to do that with xbindkeys, that's is what I've been using for a long time. But xbindkeys is X11 and I would like to switch to Wayland, so I'm trying to achieve the same with input-remapper.

--- \*\* **Edit 2** \*\* ---

I kinda got it:
![image](https://user-images.githubusercontent.com/5483864/180935394-4ee02341-e44c-43bf-9fd2-ad976947c9e0.png)

This means: if I quickly release forward button after pressing, do forward action. Otherwise I'm holding the button so don't do nothing.

This does the job, I'm fine with it, but I'll leave this interesting question: is it possible for input-remapper to only do something on input RELEASE? Example: on pressing `a` key, only print `a` after key release, no matter how long the key was pressed.

The closest I found is `if_single`, but apparently it's restricted to keyboard keys. There should be something equivalent for mouse buttons/axis.

--- \*\* **Edit 3** \*\* ---

I just started a Wayland session and here I needed to remove "Wheel HI RES" from recorded input in order to be able to cancel scrolling.

**Comment by jonasBoss** [comment]
>I just started a Wayland session and here I needed to remove "Wheel HI RES" from recorded input in order to be able to cancel scrolling.

If you map  "Wheel Hi Res" to disable it should work for both wayland and x11

> So I need to disable forward action on "button down" (hold) and enable it on "button up" (release), because I'm using it as a modifier. 

There is the `if_single` macro which does this if you don't write a `else` part: `if_single(key(a), )` will press "a" at release, and nothing if you pressed another button in the meantime.

**However:** this will not work for you because the wheel is not a button, but we can abuse the fact that `if_single` executes the `then` macro at release by using variables:

 - mapping 1: `Button EXTRA` -> `set(a, "not_scrolled").if_single( if_eq($a, "not_scrolled",key(BTN_EXTRA)),)`
 - mapping 2: `Wheel Up + Button EXTRA` -> `set(a, "scrolled").key(XF86AudioRaiseVolume)`
 - mapping 3: `Wheel HI RES +` -> `disable`

mapping 1 sets the variable `a` to `"not_scrolled"` and then `if_single` waits for the release. It then checks if `a` is still `"not_scrolled"` and presses the key or not.

**Reply by pintassilgo** [reply]
I needed to switch back to X11, reinstalled from latest beta commit and I'm unable to prevent scrolling when I'm doing the combination, could you or someone else please help me?

I need to change volume by turning the wheel while another mouse button is hold down. I will use forward button + wheel, but if you don't have mouse with extra button it can be tested as something like left click + wheel.

As said, the issue is that the scrolling isn't being prevented, volume is changed AND page scrolls (unexpectedly).

Edit: I think I got it:

  - Button EXTRA: `set(a, "not_scrolled").if_single( if_eq($a, "not_scrolled",key(BTN_EXTRA)),)`
  - Button EXTRA + Wheel Down:  `set(a, "scrolled").key(XF86AudioLowerVolume)`
  - Button EXTRA + Wheel Up:  `set(a, "scrolled").key(XF86AudioRaiseVolume)`
  - Button EXTRA + Wheel HI RES +: `disable`
  - Button EXTRA + Wheel HI RES -: `disable`
  
  `Button EXTRA` with `Target: mouse`, all others with `keyboard`.

**Reply by kevinfiol** [reply]
@pintassilgo Thanks a ton for posting this; I recently moved to Wayland and was saddened by not being able to use xbindkeys anymore. This was very helpful!
